### PR TITLE
ci: support arn automation within release pipeline

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

- Enable automation for the AWS lambda publishing.
- Create ARN table with the details in each region.
- Add two parameters:
  - whether to publish the AWS lambda, to support the old releases without lambda support.
  - build number for the release job where the AWS zip file is stored.   

## Follow ups

- This PR is the very first attempt to support the automated publishing. We will merge it and test and if it works then:
  - re-enable the disabled steps
  - test a full release 
- Upload the  `aws-lambda` zip file in the GiHub release, so far disabled but we can revisit
- Publish the ARN details somewhere else rather than in the GitHub releases

## Issues

Superseedes https://github.com/elastic/apm-agent-java/pull/2452

## Test

image

### CODE REVIEW

File: .ci/release/Jenkinsfile [hunk 1]
```diff
@@ -164,10 +167,38 @@ pipeline {
...
+        stage('Publish AWS Lambda') {
+          when {
+            expression { params.publish_aws_lambda }
+          }
+          environment {
+            SOURCE_AWS_FILE = "elastic-apm-java-aws-lambda-layer-${RELEASE_VERSION}.zip"
+          }
+          steps {
+            setEnvVar('ELASTIC_LAYER_NAME', "elastic-apm-java${env.RELEASE_AWS_LAMBDA_VERSION}")
+            withAWSEnv(secret: 'secret/observability-team/ci/service-account/apm-aws-lambda', forceInstallation: true, version: '2.4.10') {
```
Comment (v1v) [inline .ci/release/Jenkinsfile:179]: probably this requires the role_vault and so on....

File: .ci/release/Jenkinsfile [hunk 2]
```diff
@@ -164,10 +167,39 @@ pipeline {
...
+                       forceInstallation: true, version: '2.4.10') {
+              dir("${BASE_DIR}"){
+                // NOTE: copy the file from the last successful build. Therefore, this is a hard requirement
+                copyArtifacts(
+                  filter: "src/github.com/elastic/apm-agent-java/elastic-apm-agent/target/${SOURCE_AWS_FILE}",
+                  target: 'elastic-apm-agent/target',
+                  fingerprintArtifacts: true,
+                  flatten: true,
+                  projectName: 'elastic+apm-agent-java+release',
+                  selector: lastSuccessful()
```
Comment (v1v) [inline .ci/release/Jenkinsfile:189]: Enable the UI parameter to use in this selector.

File: .ci/release/Jenkinsfile [hunk 3]
```diff
@@ -164,10 +167,39 @@ pipeline {
...
+                )
+                sh(label: 'make publish-in-all-aws-regions', script: 'make -C .ci publish-in-all-aws-regions')
+                sh(label: 'make create-arn-file', script: 'make -C .ci create-arn-file')
```
Comment (v1v) [inline .ci/release/Jenkinsfile:192]: Archive the file to see the look%feel

File: .ci/create-arn-table.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+# Create the AWS ARN table given the below environment variables:
+#
+#   - AWS_FOLDER      - that's the location of the publish-layer-version output for each region
+#	- SUFFIX_ARN_FILE - that's the output file.
+#
+
+ARN_FILE=${SUFFIX_ARN_FILE}
+
+{
+	echo "### ARN"
```
Comment (AlexanderWert) [inline .ci/create-arn-table.sh:14]:
```suggestion
	echo "### ARNs of the APM Java Agent's AWS Lambda Layer"
```

File: .ci/Makefile [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+		AWS_DEFAULT_REGION="$${AWS_DEFAULT_REGION}" ELASTIC_LAYER_NAME=$(ELASTIC_LAYER_NAME) $(MAKE) grant-public-layer-access; \
+	done <.regions
+
+# Publish the given LAYER in the given AWS region
+publish: validate-layer-name validate-aws-default-region
+	@aws lambda \
+		--output json \
+		publish-layer-version \
+		--layer-name "$(ELASTIC_LAYER_NAME)" \
+		--description "AWS Lambda Extension Layer for Elastic APM Java" \
```
Comment (AlexanderWert) [inline .ci/Makefile:36]:
```suggestion
		--description "AWS Lambda Extension Layer for the Elastic APM Java Agent" \
```

Comment (AlexanderWert) [reply .ci/Makefile:36]: ⬆️ I think for the Java agent layer we should also set the compatible runtimes option when publishing the layer.

Comment (v1v) [reply .ci/Makefile:36]: image

For the record -> [here](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html)

Good catch! Thanks 🙇

File: .ci/release/Jenkinsfile [hunk 6]
```diff
@@ -164,10 +189,47 @@ pipeline {
...
+              dir("${BASE_DIR}"){
+                // NOTE: copy the file from the last successful build. Therefore, this is a hard requirement
+                copyArtifacts(
+                  filter: "src/github.com/elastic/apm-agent-java/elastic-apm-agent/target/${SOURCE_AWS_FILE}",
+                  target: 'elastic-apm-agent/target',
+                  fingerprintArtifacts: true,
+                  flatten: true,
+                  projectName: 'elastic+apm-agent-java+release',
+                  selector: specific(params.release_build)
+                )
```
Comment (v1v) [inline .ci/release/Jenkinsfile:212]: image

File: .ci/release/Jenkinsfile [hunk 7]
```diff
@@ -164,10 +188,42 @@ pipeline {
...
+          }
+          steps {
+            setEnvVar('ELASTIC_LAYER_NAME', "elastic-apm-java${env.RELEASE_AWS_LAMBDA_VERSION}")
+            withAWSEnv(secret: 'secret/observability-team/ci/service-account/apm-aws-lambda', role_id: 'apm-vault-role-id', secret_id: 'apm-vault-secret-id',
+                       forceInstallation: true, version: '2.4.10') {
+              dir("${BASE_DIR}"){
+                dir ('elastic-apm-agent/target') {
+                  // TODO: copy file from a google bucket
+                  sh(label: 'fetch AWS lambda file', script: "wget https://github.com/elastic/apm-agent-java/releases/download/v${RELEASE_VERSION}/${SOURCE_AWS_FILE} -O ${SOURCE_AWS_FILE}")
+                }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:206]: This should be changed!

Comment (v1v) [reply .ci/release/Jenkinsfile:206]: https://github.com/elastic/apm-agent-java/pull/2453#discussion_r804624611

File: .ci/create-arn-table.sh [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	echo '|------|---|'
+} > "${ARN_FILE}"
+
+for f in $(ls "${AWS_FOLDER}"); do
+	LAYER_VERSION_ARN=$(grep '"LayerVersionArn"' "$AWS_FOLDER/${f}" | cut -d":" -f2- | sed 's/ //g' | sed 's/"//g' | cut -d"," -f1)
+	echo "INFO: create-arn-table ARN(${LAYER_VERSION_ARN}):region(${f}))"
+	echo "|${f}|${LAYER_VERSION_ARN}|" >> "${ARN_FILE}"
+done
+
+echo '' >> "${ARN_FILE}"
```
Comment (cachedout) [inline .ci/create-arn-table.sh:26]: A `trap` here so this cleanup always runs, perhaps?

File: .ci/release/Jenkinsfile [hunk 9]
```diff
@@ -16,6 +16,9 @@ pipeline {
     SLACK_CHANNEL = '#apm-agent-java'
     NOTIFY_TO = 'build-apm+apm-agent-java@elastic.co'
     BRANCH_SPECIFIER = "${params.branch_specifier}"
+    SUFFIX_ARN_FILE = 'arn-file.md'
+    RELEASE_AWS_LAMBDA_VERSION = '-ver-1-29-0'
+    RELEASE_VERSION = '1.29.0'
```
Comment (v1v) [inline .ci/release/Jenkinsfile:21]:
```suggestion
    //RELEASE_AWS_LAMBDA_VERSION = '-ver-1-29-0'
    //RELEASE_VERSION = '1.29.0'
```

File: .ci/release/Jenkinsfile [hunk 10]
```diff
@@ -63,6 +67,9 @@ pipeline {
       options { skipDefaultCheckout () }
       stages{
         stage('Check oss.sonatype.org') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:72]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 11]
```diff
@@ -80,6 +87,7 @@ pipeline {
             allOf {
               expression { params.check_branch_ci_status }
               expression { env.BRANCH_SPECIFIER != 'stable' }
+              expression { return false }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:90]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 12]
```diff
@@ -92,6 +100,9 @@ pipeline {
           }
         }
         stage('Require confirmation that CHANGELOG.asciidoc has been updated') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:105]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 13]
```diff
@@ -107,6 +118,9 @@ pipeline {
           }
         }
         stage('Set release version') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:123]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 14]
```diff
@@ -131,25 +145,35 @@ pipeline {
...
                 env.BRANCH_DOT_X = user_release_version.substring(0, user_release_version.indexOf('.'))+'.x'
+                env.RELEASE_AWS_LAMBDA_VERSION = '-ver-' + user_release_version.replaceAll('\\.', '-')
               }
             }
           }
         }
         stage('Wait on internal CI') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:156]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 15]
```diff
@@ -131,25 +145,35 @@ pipeline {
...
           steps {
             notifyStatus(slackStatus: 'warning', subject: "[${env.REPO}] Release ready to be pushed",
                           body: "Please go to (<${env.BUILD_URL}input|here>) to approve or reject within 12 hours.")
             input(message: "Start the release job on the internal CI. Click 'Proceed' once the job has succeeded or click 'Abort' if the release has failed and then manually undo the release.")
           }
         }
         stage('Nexus release') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:166]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 16]
```diff
@@ -131,25 +145,35 @@ pipeline {
...
           steps {
             notifyStatus(slackStatus: 'warning', subject: "[${env.REPO}] Release ready to be published in Nexus",
                          body: "Please go to (<https://oss.sonatype.org/|here>) to proceed with the manual nexus release. Login details in LastPass")
             input(message: "Go to https://oss.sonatype.org and proceed with the steps to close and release the staging artifact.")
           }
         }
         stage('Major Branch create/update') {
+          when {
+            expression { return false }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:176]:
```suggestion
```

File: .ci/release/Jenkinsfile [hunk 17]
```diff
@@ -164,10 +188,42 @@ pipeline {
             }
           }
         }
+        stage('Publish AWS Lambda') {
+          when {
+            expression { params.publish_aws_lambda }
+          }
```
Comment (v1v) [inline .ci/release/Jenkinsfile:194]:
```suggestion
          when {
            //expression { params.publish_aws_lambda }
            expression { return false }
          }
```

**Comment by v1v** [comment]
In order to test this we need to merge it, and these changes are not validated in the apm-ci as they are not executed as part of the main pipeline. I'll proceed and merge to be able to move forward with the release

**Pull request merged**